### PR TITLE
Re-implement ia2.

### PR DESCRIPTION
**Pull request by NullHypothesis**
This PR improves the design of ia2 (and renames it to tokenizer) by breaking it into four components: receivers (which receive incoming data), tokenizers (which turn data into tokens), forwarders (which send data somewhere), and aggregators (which aggregate and/or augment data).

### CODE REVIEW

File: aggregator_addr_msg_test.go [hunk 1]
```diff
@@ -1,21 +1,30 @@
...
+	o := ourString(s)
+	if !bytes.Equal(o.bytes(), []byte(s)) {
+		t.Fatalf("Expected %s but got %s.", []byte(s), o.bytes())
+	}
+}
+
+func TestSerialization(t *testing.T) {
 	walletID := uuid.NewV4()
-	keyID := KeyID{UUID: uuid.NewV4()}
+	kID := keyID{UUID: uuid.NewV4()}
```
Comment (rillian) [inline aggregator_addr_msg_test.go:22]: satori/go.uuid v1.2.0 has a [randomness bug](https://nvd.nist.gov/vuln/detail/CVE-2021-3538). It's fixed in the repo, but doesn't seem to have released an update. Can you use a newer git revision, or switch to a more recently-maintained version? google/uuid seems to be almost as popular.

File: main.go [hunk 2]
```diff
@@ -1,80 +1,166 @@
...
-	// anonymize IP addresses.  Once the key expires, we rotate it by
-	// generating a new one.
-	KeyExpiration = time.Hour * 24 * 30 * 6
-	// localProxy determines the IP address and port of the enclave-internal
-	// proxy that translates between AF_INET and AF_VSOCK.
-	localProxy = "127.0.0.1:1080"
+	defaultTokenizer  = tokenizerHmac
+	defaultAggregator = aggregatorSimple
+	defaultReceiver   = receiverStdin
+	defaultForwarder  = forwarderStdout
```
Comment (rillian) [inline main.go:32]: Can these be in the same order as the specific constants above?

Comment (NullHypothesis) [reply main.go]: Fixed in 7a07cb7.

File: aggregator_addr_msg_test.go (hunk 1 above)
Comment (NullHypothesis) [reply aggregator_addr_msg_test.go]: Yikes. Looks like our current dependency is unmaintained. I replaced it with Google's in a445989.

File: receiver_web.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	w.router = newRouter(w.in)
+
+	return w
+}
+
+func newRouter(inbox chan serializer) *chi.Mux {
+	r := chi.NewRouter()
+	r.Get("/v1/confirmation/token/{walletID}", getConfTokenHandler(inbox))
+	r.Get("/v2/confirmation/token/{walletID}", getConfTokenHandler(inbox))
```
Comment (rillian) [inline receiver_web.go:58]: The two api versions are identical? Should there be tests for the `/v1/**` paths as well?

File: tokenizer_cryptopan.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	return &cryptoPAnTokenizer{}
+}
+
+func (c *cryptoPAnTokenizer) isBlobSupported(b []byte) bool {
+	return len(b) == ipv4Len || len(b) == ipv6Len
+}
+
+func (c *cryptoPAnTokenizer) tokenize(s serializer) (token, error) {
+	if len(c.key) == 0 {
+		return nil, errNoKey
```
Comment (rillian) [inline tokenizer_cryptopan.go:39]: :heart:

**Review by rillian — APPROVED** [review]
Mostly questions for my own understanding. Otherwise I think it's fine to merge. Thanks for your patience.

File: aggregator_addr.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		for {
+			select {
+			case <-a.done:
+				return
+			case <-fwdTicker.C:
+				if err := a.flush(); err != nil {
+					l.Printf("Failed to forward addresses: %v", err)
+				}
+			case <-keyTicker.C:
+				if err := a.tokenizer.resetKey(); err != nil {
```
[the remark this answers is not in the document]
Comment (NullHypothesis) [reply aggregator_addr.go:111]: You are right, it's best to fail. Fixed in e09148c.

File: aggregator_addr.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+				}
+			}
+		}
+	}()
+}
+
+// stop stops the address aggregator.
+func (a *addrAggregator) stop() {
+	close(a.done)
+	a.wg.Wait()
```
[the remark this answers is not in the document]
Comment (NullHypothesis) [reply aggregator_addr.go:140]: Yes, WaitGroups are thread-safe.

File: aggregator_addr.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	l.Println("Stopped address aggregator.")
+}
+
+// processRequest processes an incoming client request.
+func (a *addrAggregator) processRequest(req *clientRequest) error {
+	a.Lock()
+	defer a.Unlock()
```
[the remark this answers is not in the document]
Comment (NullHypothesis) [reply aggregator_addr.go:147]: This lock protects `a.addrs` which keeps track of our wallet ID/IP address mappings. That said, the tokenizers need locking too, which I somehow missed. Fixed in 2a4125b.

File: receiver_web.go (hunk 3 above)
Comment (NullHypothesis) [reply receiver_web.go]: Yes, they were identical but I was just told that we retired the v1 endpoint several months ago. Fixed in d9dcd89. It turns out that there's already a v3 endpoint that's used in staging. I created issue https://github.com/brave/ia2/issues/23 for this.

File: aggregator_addr.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (rillian) [reply aggregator_addr.go:147]: Ok, the tokenizers do have internal state (or at least rotate their keys) and the aggregator mutex doesn't cover all the calls; in particular `addrAggregator::start` calls both `resetKey` and `tokenize` directly.

I was thinking that if the tokenizer was thread-safe you could not take the lock until right before you access `a.addrs`, reducing contention when you're getting lots of bad requests. But the way you have it protects against the someone calling `use` to change the tokenizer in-between the two calls, which is the worthwhile.

**Review by rillian — APPROVED** [review]

**Review by rillian — DISMISSED** [review]

**Review by rillian — APPROVED** [review]

**Pull request merged**